File: src/main/java/me/senseiwells/arucas/utils/impl/ArucasList.java
package me.senseiwells.arucas.utils.impl;

import me.senseiwells.arucas.api.ISyntax;
import me.senseiwells.arucas.extensions.util.LazyValue;
import me.senseiwells.arucas.throwables.BuiltInException;
import me.senseiwells.arucas.throwables.CodeError;
import me.senseiwells.arucas.utils.Context;
import me.senseiwells.arucas.utils.LazyGetter;
import me.senseiwells.arucas.utils.StringUtils;
import me.senseiwells.arucas.values.Value;
import me.senseiwells.arucas.values.ValueIdentifier;

import java.util.*;

/**
 * Custom list implementation.
 * Most of the code was taken from ArrayList,
 * this implements {@link me.senseiwells.arucas.values.ValueIdentifier}
 * as it is easier to implement these
 * methods natively.
 */
public class ArucasList implements IArucasCollection, List<Value>, ValueIdentifier {
	private static final Object DEADLOCKED_HANDLER = new Object();
	private static final Value[] DEFAULT_DATA = { };
	private static final int DEFAULT_CAPACITY = 10;

	private Value[] valueData;
	private int size;

	public ArucasList() {
		this.valueData = DEFAULT_DATA;
	}

	public ArucasList(ArucasList valueList) {
		Value[] valueArray = valueList.toArray();
		this.size = valueArray.length;
		this.valueData = this.size == 0 ? DEFAULT_DATA : valueArray;
	}

	private ArucasList(Value[] valueData) {
		this.size = valueData.length;
		this.valueData = this.size == 0 ? DEFAULT_DATA : valueData;
	}

	@Override
	public synchronized int size() {
		return this.size;
	}

	@Override
	public String getAsStringSafe() {
		return "<list>";
	}

	@Override
	public synchronized boolean isEmpty() {
		return this.size == 0;
	}

	@Override
	public synchronized Value get(int index) {
		this.checkExistingIndex(index);
		return this.valueData[index];
	}

	public synchronized boolean contains(Context context, Value value) throws CodeError {
		return this.indexOf(context, value) >= 0;
	}

	public synchronized boolean containsAll(Context context, Collection<? extends Value> valueList) throws CodeError {
		if (this.size < valueList.size()) {
			return false;
		}

		if (valueList instanceof ArucasList arucasList) {
			// This calls .toArray() because it may cause deadlocks
			valueList = Arrays.asList(arucasList.toArray());
		}

		for (Value value : valueList) {
			if (!this.contains(context, value)) {
				return false;
			}
		}
		return true;
	}

	public synchronized int indexOf(Context context, Value value) throws CodeError {
		Value[] valueData = this.valueData;
		for (int i = 0; i < this.size; i++) {
			if (value.isEquals(context, valueData[i])) {
				return i;
			}
		}
		return -1;
	}

	@Override
	public Value set(int index, Value element) {
		this.checkExistingIndex(index);
		Value oldValue = this.valueData[index];
		this.valueData[index] = element;
		return oldValue;
	}

	@Override
	public synchronized boolean add(Value value) {
		this.add(value, this.valueData, this.size);
		return true;
	}

	@Override
	public synchronized void add(int index, Value value) {
		this.checkAddIndex(index);
		final int size = this.size;
		Value[] valueData = this.valueData;
		if (size == valueData.length) {
			valueData = this.grow();
		}
		System.arraycopy(valueData, index, valueData, index + 1, size - index);
		valueData[index] = value;
		this.size = size + 1;
	}

	private synchronized void add(Value value, Value[] valueData, int size) {
		if (size == valueData.length) {
			valueData = this.grow();
		}
		valueData[size] = value;
		this.size = size + 1;
	}

	@Override
	public synchronized boolean addAll(Collection<? extends Value> values) {
		if (values instanceof ArucasList arucasList) {
			return this.addAll(arucasList.toArray());
		}
		return this.addAll(values.toArray(Value[]::new));
	}


	@Override
	public boolean addAll(int index, Collection<? extends Value> c) {
		this.checkAddIndex(index);

		Value[] valueData = c instanceof ArucasList arucasList ? arucasList.toArray() : c.toArray(Value[]::new);
		int numNew = valueData.length;
		if (numNew == 0) {
			return false;
		}

		Value[] elementData = this.valueData;
		final int size = this.size;
		if (numNew > elementData.length - size) {
			elementData = this.grow(size + numNew);
		}

		int numMoved = size - index;
		if (numMoved > 0) {
			System.arraycopy(elementData, index, elementData, index + numNew, numMoved);
		}
		System.arraycopy(valueData, 0, elementData, index, numNew);
		this.size = size + numNew;
		return true;
	}

	private synchronized boolean addAll(Value[] valueArray) {
		int newSize = valueArray.length;
		if (newSize == 0) {
			return false;
		}
		Value[] valueData = this.valueData;
		final int size = this.size;
		if (newSize > valueData.length - size) {
			valueData = this.grow(size + newSize);
		}
		System.arraycopy(valueArray, 0, valueData, size, newSize);
		this.size = size + newSize;
		return true;
	}

	@Override
	public synchronized Value remove(int index) {
		this.checkAddIndex(index);
		final Value[] valueData = this.valueData;
		Value oldValue = valueData[index];
		this.remove(valueData, index);
		return oldValue;
	}

	public synchronized boolean remove(Context context, Value value) throws CodeError {
		final Value[] valueData = this.valueData;
		for (int i = 0; i < this.size; i++) {
			if (value.isEquals(context, valueData[i])) {
				this.remove(valueData, i);
				return true;
			}
		}
		return false;
	}

	private synchronized void remove(Value[] valueData, int index) {
		final int newSize = this.size - 1;
		if (newSize > index) {
			System.arraycopy(valueData, index + 1, valueData, index, newSize - index);
		}
		valueData[this.size = newSize] = null;
	}

	@Override
	public boolean removeAll(Collection<?> collection) {
		return this.batchRemove(collection, this.size);
	}

	private boolean batchRemove(Collection<?> collection, final int to) {
		final Value[] valueData = this.valueData;
		int i = 0;
		for (; ; i++) {
			if (i == to) {
				return false;
			}
			if (collection.contains(valueData[i])) {
				break;
			}
		}
		int j = i++;
		for (Value value; i < to; i++) {
			if (!collection.contains(value = valueData[i])) {
				valueData[j++] = value;
			}
		}
		this.shiftTailOverGap(valueData, j, to);
		return true;
	}

	@Override
	public synchronized void clear() {
		final Value[] valueData = this.valueData;
		for (int to = this.size, i = this.size = 0; i < to; i++) {
			valueData[i] = null;
		}
	}

	@Override
	public Value[] toArray() {
		return Arrays.copyOf(this.valueData, this.size);
	}

	@Override
	public <T> T[] toArray(T[] a) {
		if (a instanceof Value[]) {
			@SuppressWarnings("unchecked")
			T[] values = (T[]) this.toArray();
			return values;
		}
		throw new ArrayStoreException();
	}

	@Override
	public synchronized ArucasList subList(int fromIndex, int toIndex) {
		return new ArucasList(Arrays.copyOfRange(this.valueData, fromIndex, toIndex));
	}

	private synchronized Value[] grow() {
		return this.grow(this.size + 1);
	}

	private synchronized Value[] grow(int minCapacity) {
		final int oldCapacity = this.valueData.length;
		if (oldCapacity > 0 || this.valueData != DEFAULT_DATA) {
			int newCapacity = newLength(oldCapacity, minCapacity - oldCapacity, oldCapacity >> 1);
			return this.valueData = Arrays.copyOf(this.valueData, newCapacity);
		}
		return this.valueData = new Value[Math.max(DEFAULT_CAPACITY, minCapacity)];
	}

	private synchronized void shiftTailOverGap(Value[] valueData, int low, int high) {
		System.arraycopy(valueData, high, valueData, low, this.size - high);
		for (int to = this.size, i = (this.size -= high - low); i < to; i++) {
			valueData[i] = null;
		}
	}

	private synchronized void checkAddIndex(int index) {
		if (index < 0 || index > this.size) {
			throw new BuiltInException("Index out of bounds");
		}
	}

	private synchronized void checkExistingIndex(int index) {
		if (index < 0 || index >= this.size) {
			throw new BuiltInException("Index out of bounds");
		}
	}

	@Override
	public synchronized int getHashCode(Context context) throws CodeError {
		final Value[] valueData = this.valueData;
		int hashCode = 1;
		for (int i = 0; i < this.size; i++) {
			Value value = valueData[i];
			hashCode = 31 * hashCode + (value == null ? 0 : value.getHashCode(context));
		}
		return hashCode;
	}

	@Override
	public boolean isEquals(Context context, Value other) throws CodeError {
		if (!(other.getValue() instanceof ArucasList that)) {
			return false;
		}

		if (this == that) {
			return true;
		}

		synchronized (DEADLOCKED_HANDLER) {
			synchronized (this) {
				synchronized (that) {
					if (this.size != that.size) {
						return false;
					}

					for (int i = 0; i < this.size; i++) {
						if (!this.valueData[i].isEquals(context, that.valueData[i])) {
							return false;
						}
					}
				}
			}
		}

		return true;
	}

	@Override
	public String getAsStringUnsafe(Context context, ISyntax position) throws CodeError {
		return "[" + IArucasCollection.super.getAsStringUnsafe(context, position) + "]";
	}

	@Override
	public Collection<? extends Value> asCollection() {
		return this;
	}

	@Override
	public synchronized String getAsString(Context context) throws CodeError {
		if (this.isEmpty()) {
			return "[]";
		}

		StringBuilder stringBuilder = new StringBuilder();
		stringBuilder.append("[");

		int i = 0;
		while (i < this.size) {
			stringBuilder.append(StringUtils.toPlainString(context, this.valueData[i]));
			if (++i < this.size) {
				stringBuilder.append(", ");
			}
		}

		stringBuilder.append("]");
		return stringBuilder.toString();
	}

	@Override
	public synchronized Iterator<Value> iterator() {
		return new Iterator<>() {
			int cursor = 0;

			@Override
			public synchronized boolean hasNext() {
				return this.cursor != ArucasList.this.size;
			}

			@Override
			public synchronized Value next() {
				final int i = this.cursor;
				if (i >= ArucasList.this.size) {
					throw new NoSuchElementException();
				}
				Value[] valueData = ArucasList.this.valueData;
				this.cursor = i + 1;
				return valueData[i];
			}
		};
	}

	/**
	 * These methods should not be used instead
	 * {@link #getAsString(Context)},
	 * {@link #getHashCode(Context)},
	 * {@link #isEquals(Context, Value)},
	 * should be used.
	 */

	@Deprecated
	@Override
	public final int hashCode() {
		return super.hashCode();
	}

	@Deprecated
	@Override
	public final boolean equals(Object obj) {
		return this == obj;
	}

	@Deprecated
	@Override
	public final String toString() {
		return super.toString();
	}

	/**
	 * These methods are unsupported.
	 */

	@Override
	public boolean contains(Object o) { throw new UnsupportedOperationException(); }

	@Override
	public boolean remove(Object o) { throw new UnsupportedOperationException(); }

	@Override
	public boolean containsAll(Collection<?> c) { throw new UnsupportedOperationException(); }

	@Override
	public boolean retainAll(Collection<?> c) { throw new UnsupportedOperationException(); }

	@Override
	public int indexOf(Object o) { throw new UnsupportedOperationException(); }

	@Override
	public int lastIndexOf(Object o) { throw new UnsupportedOperationException(); }

	@Override
	public ListIterator<Value> listIterator() { throw new UnsupportedOperationException(); }

	@Override
	public ListIterator<Value> listIterator(int index) { throw new UnsupportedOperationException(); }


	public static ArucasList of(Value... values) {
		return new ArucasList(values);
	}

	public static ArucasList ofLazy(LazyGetter... getters) {
		ArucasList list = new ArucasList();
		for (LazyGetter getter : getters) {
			list.add(LazyValue.of(getter));
		}
		return list;
	}

	public static List<Value> arrayListOf(Value... values) {
		return new ArrayList<>(Arrays.asList(values));
	}

	public static final int MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;

	// Taken from ArraysSupport
	public static int newLength(int oldLength, int minGrowth, int prefGrowth) {
		int newLength = Math.max(minGrowth, prefGrowth) + oldLength;
		if (newLength - MAX_ARRAY_LENGTH <= 0) {
			return newLength;
		}
		return hugeLength(oldLength, minGrowth);
	}

	// Taken from ArraysSupport
	private static int hugeLength(int oldLength, int minGrowth) {
		int minLength = oldLength + minGrowth;
		if (minLength < 0) { // overflow
			throw new OutOfMemoryError("Required array length too large");
		}
		if (minLength <= MAX_ARRAY_LENGTH) {
			return MAX_ARRAY_LENGTH;
		}
		return Integer.MAX_VALUE;
	}
}

File: src/main/kotlin/me/senseiwells/arucas/utils/impl/ArucasList.kt
package me.senseiwells.arucas.utils.impl

import me.senseiwells.arucas.classes.ClassInstance
import me.senseiwells.arucas.core.Interpreter
import me.senseiwells.arucas.exceptions.runtimeError
import kotlin.math.max

class ArucasList private constructor(
    @Volatile var data: Array<ClassInstance?>,
): MutableList<ClassInstance>, ArucasCollection {
    companion object {
        private const val MAX_ARRAY_LENGTH = Int.MAX_VALUE - 8
        private const val DEFAULT_CAPACITY = 10

        private val DEFAULT_DATA = emptyArray<ClassInstance?>()

        @JvmStatic
        fun of(vararg elements: ClassInstance): ArucasList {
            return ArucasList(arrayOf(*elements)).also {
                it.size = elements.size
            }
        }
    }

    @Volatile
    override var size: Int = 0

    constructor(): this(DEFAULT_DATA)

    constructor(list: ArucasList): this(list.toArrayNullable()) {
        this.size = this.data.size
    }

    constructor(collection: Collection<ClassInstance>): this(collection.toTypedArray()) {
        this.size = this.data.size
    }

    fun toArray(): Array<ClassInstance> = Array(this.size) { this.data[it]!! }

    private fun toArrayNullable(): Array<ClassInstance?> = Array(this.size) { this.data[it] }

    override fun length() = this.size

    override fun asCollection() = this

    @Synchronized
    fun contains(interpreter: Interpreter, element: ClassInstance): Boolean {
        return this.indexOf(interpreter, element) >= 0
    }

    @Synchronized
    fun containsAll(interpreter: Interpreter, elements: ArucasCollection): Boolean {
        if (this.size < elements.length()) {
            return false
        }

        var mutElements: Iterable<ClassInstance> = elements
        if (mutElements is ArucasList) {
            // Avoiding deadlocks
            mutElements = mutElements.toArray().asList()
        }

        for (element in mutElements) {
            if (!this.contains(interpreter, element)) {
                return false
            }
        }
        return true
    }

    @Synchronized
    override fun get(index: Int): ClassInstance {
        this.checkExistingIndex(index)
        return this.data[index]!!
    }

    @Synchronized
    fun indexOf(interpreter: Interpreter, element: ClassInstance): Int {
        for (i in 0 until this.size) {
            if (element.equals(interpreter, this[i])) {
                return i
            }
        }
        return -1
    }

    @Synchronized
    override fun isEmpty(): Boolean {
        return this.size == 0
    }

    @Synchronized
    override fun iterator(): MutableIterator<ClassInstance> {
        return this.listIterator()
    }

    @Synchronized
    fun lastIndexOf(interpreter: Interpreter, element: ClassInstance): Int {
        for (i in (this.size - 1) downTo 0) {
            if (element.equals(interpreter, this[i])) {
                return i
            }
        }
        return -1
    }

    @Synchronized
    override fun add(element: ClassInstance): Boolean {
        this.add(this.size, element, this.data)
        return true
    }

    @Synchronized
    override fun add(index: Int, element: ClassInstance) {
        this.checkAddIndex(index)
        val size = this.size
        var data = this.data
        if (size == data.size) {
            data = this.grow()
        }
        System.arraycopy(data, index, data, index + 1, size - index)
        data[index] = element
        this.size = size + 1
    }

    @Synchronized
    override fun addAll(index: Int, elements: Collection<ClassInstance>): Boolean {
        this.checkAddIndex(index)

        val data = ArucasCollection.safeCollection(elements)
        val expansion = data.size
        if (expansion == 0) {
            return false
        }

        var thisData = this.data
        val size = this.size
        if (expansion > thisData.size - size) {
            thisData = this.grow(size + expansion)
        }

        val moved = size - index
        if (moved > 0) {
            System.arraycopy(thisData, index, thisData, index + expansion, moved)
        }
        System.arraycopy(data, 0, thisData, index, expansion)
        this.size = size + expansion
        return true
    }

    @Synchronized
    override fun addAll(elements: Collection<ClassInstance>): Boolean {
        return this.addAll(ArucasCollection.safeCollection(elements))
    }

    @Synchronized
    override fun clear() {
        val data = this.data
        val size = this.size
        for (i in 0 until size) {
            data[i] = null
        }
        this.size = 0
    }

    @Synchronized
    override fun listIterator(): MutableListIterator<ClassInstance> {
        return this.ArucasIterator()
    }

    @Synchronized
    override fun listIterator(index: Int): MutableListIterator<ClassInstance> {
        return this.ArucasIterator(index)
    }

    @Synchronized
    fun remove(interpreter: Interpreter, element: ClassInstance): Boolean {
        val data = this.data
        for (i in 0 until this.size) {
            if (element.equals(interpreter, data[i]!!)) {
                this.remove(data, i)
                return true
            }
        }
        return false
    }

    @Synchronized
    override fun removeAll(elements: Collection<ClassInstance>): Boolean {
        return this.batchRemove(elements, false)
    }

    @Synchronized
    override fun removeAt(index: Int): ClassInstance {
        this.checkAddIndex(index)
        val valueData: Array<ClassInstance?> = this.data
        val oldValue: ClassInstance = valueData[index]!!
        this.remove(valueData, index)
        return oldValue
    }

    @Synchronized
    override fun retainAll(elements: Collection<ClassInstance>): Boolean {
        return this.batchRemove(elements, true)
    }

    @Synchronized
    override fun set(index: Int, element: ClassInstance): ClassInstance {
        this.checkExistingIndex(index)
        val old = this.data[index]!!
        this.data[index] = element
        return old
    }

    @Synchronized
    override fun subList(fromIndex: Int, toIndex: Int): ArucasList {
        return ArucasList(this.data.copyOfRange(fromIndex, toIndex))
    }

    @Synchronized
    private fun add(size: Int, element: ClassInstance, data: Array<ClassInstance?>) {
        var mutData = data
        if (size == data.size) {
            mutData = this.grow()
        }
        mutData[size] = element
        this.size = size + 1
    }

    @Synchronized
    private fun addAll(elements: Array<ClassInstance>): Boolean {
        val expansion = elements.size
        if (expansion == 0) {
            return false
        }

        var data: Array<ClassInstance?> = this.data
        val size = this.size
        if (expansion > data.size - size) {
            data = this.grow(size + expansion)
        }

        System.arraycopy(elements, 0, data, size, expansion)
        this.size = size + expansion
        return true
    }

    @Synchronized
    private fun remove(elements: Array<ClassInstance?>, index: Int) {
        val newSize = this.size - 1
        if (newSize > index) {
            System.arraycopy(elements, index + 1, elements, index, newSize - index)
        }
        elements[newSize.also { this.size = it }] = null
    }

    @Synchronized
    private fun batchRemove(collection: Collection<ClassInstance>, compliment: Boolean): Boolean {
        val elements = ArucasCollection.safeCollection(collection)
        val data = this.data
        val to = this.size
        var i = 0
        while (true) {
            if (i == to) {
                return false
            }
            if (data[i] in elements != compliment) {
                break
            }
            i++
        }
        var j = i++
        var instance: ClassInstance?
        while (i < to) {
            if (data[i].also { instance = it } in elements == compliment) {
                data[j++] = instance
            }
            i++
        }
        this.shiftTailOverGap(data, j, to)
        return true
    }

    @Synchronized
    private fun shiftTailOverGap(valueData: Array<ClassInstance?>, low: Int, high: Int) {
        System.arraycopy(valueData, high, valueData, low, this.size - high)
        var i = high - low.let { this.size -= it; this.size }
        while (i < this.size) {
            valueData[i++] = null
        }
    }

    @Synchronized
    private fun checkAddIndex(index: Int) {
        if (index < 0 || index > this.size) {
            runtimeError("Index out of bounds")
        }
    }

    @Synchronized
    private fun checkExistingIndex(index: Int) {
        if (index < 0 || index >= this.size) {
            runtimeError("Index out of bounds")
        }
    }

    @Synchronized
    private fun grow(): Array<ClassInstance?> {
        return this.grow(this.size + 1)
    }

    @Synchronized
    private fun grow(minCapacity: Int): Array<ClassInstance?> {
        val oldCapacity: Int = this.data.size
        if (oldCapacity > 0 || this.data !== DEFAULT_DATA) {
            val newCapacity: Int = this.newLength(oldCapacity, minCapacity - oldCapacity, oldCapacity shr 1)
            return this.data.copyOf(newCapacity).also { this.data = it }
        }
        return arrayOfNulls<ClassInstance>(max(DEFAULT_CAPACITY, minCapacity)).also { this.data = it }
    }

    private fun newLength(oldLength: Int, minGrowth: Int, prefGrowth: Int): Int {
        val newLength = minGrowth.coerceAtLeast(prefGrowth) + oldLength
        return if (newLength - MAX_ARRAY_LENGTH <= 0) newLength else this.hugeLength(oldLength, minGrowth)
    }

    private fun hugeLength(oldLength: Int, minGrowth: Int): Int {
        val minLength = oldLength + minGrowth
        if (minLength < 0) {
            throw OutOfMemoryError("Required array length too large")
        }
        return if (minLength <= MAX_ARRAY_LENGTH) MAX_ARRAY_LENGTH else Int.MAX_VALUE
    }

    @Synchronized
    fun hashCode(interpreter: Interpreter): Int {
        val data: Array<ClassInstance?> = this.data
        var hashCode = 1
        for (i in 0 until this.size) {
            val element = data[i]
            hashCode = 31 * hashCode + (element?.hashCode(interpreter) ?: 0)
        }
        return hashCode
    }

    @Synchronized
    fun equals(interpreter: Interpreter, other: ArucasList): Boolean {
        if (this === other) {
            return true
        }
        // We need to make a copy or deadlocks may occur
        val copy = other.toArray()
        if (this.size != copy.size) {
            return false
        }
        for (i in 0 until this.size) {
            if (!this[i].equals(interpreter, copy[i])) {
                return false
            }
        }
        return true
    }

    override fun toSafeString() = "<list>"

    override fun formatCollection(elements: String): String {
        return "[$elements]"
    }

    @Deprecated("Need interpreter for this operation", replaceWith = ReplaceWith("list.hashCode(interpreter)"))
    override fun hashCode(): Int {
        return System.identityHashCode(this)
    }

    @Deprecated("Need interpreter for this operation", replaceWith = ReplaceWith("list.equals(interpreter, other)"))
    override fun equals(other: Any?): Boolean {
        return this === other
    }

    @Deprecated("Need interpreter for this operation", replaceWith = ReplaceWith("list.toString(interpreter)"))
    override fun toString(): String {
        return this.toSafeString()
    }

    @Deprecated("Need interpreter for this operation", replaceWith = ReplaceWith("list.contains(interpreter, element)"))
    override fun contains(element: ClassInstance): Boolean {
        this.needInterpreterForThis()
    }

    @Deprecated("Need interpreter for this operation", replaceWith = ReplaceWith("list.containsAll(interpreter, collection)"))
    override fun containsAll(elements: Collection<ClassInstance>): Boolean {
        this.needInterpreterForThis()
    }

    @Deprecated("Need interpreter for this operation", replaceWith = ReplaceWith("list.indexOf(interpreter, element)"))
    override fun indexOf(element: ClassInstance): Int {
        this.needInterpreterForThis()
    }

    @Deprecated("Need interpreter for this operation", replaceWith = ReplaceWith("list.lastIndexOf(interpreter, element)"))
    override fun lastIndexOf(element: ClassInstance): Int {
        this.needInterpreterForThis()
    }

    @Deprecated("Need interpreter for this operation", replaceWith = ReplaceWith("list.remove(interpreter, element)"))
    override fun remove(element: ClassInstance): Boolean {
        this.needInterpreterForThis()
    }

    private inner class ArucasIterator(var cursor: Int = 0): MutableListIterator<ClassInstance> {
        var last = -1

        override fun hasPrevious(): Boolean {
            synchronized(this@ArucasList) {
                return this.cursor != 0
            }
        }

        override fun nextIndex(): Int {
            synchronized(this@ArucasList) {
                return this.cursor
            }
        }

        override fun previous(): ClassInstance {
            synchronized(this@ArucasList) {
                val i = this.cursor - 1
                if (i < 0) {
                    runtimeError("No such element")
                }
                val data = this@ArucasList.data
                this.cursor = i
                this.last = i
                return data[i]!!
            }
        }

        override fun previousIndex(): Int {
            synchronized(this@ArucasList) {
                return this.cursor - 1
            }
        }

        override fun add(element: ClassInstance) {
            synchronized(this@ArucasList) {
                val i = this.cursor
                this@ArucasList.add(i, element)
                this.cursor = i + 1
                this.last = -1
            }
        }

        override fun hasNext(): Boolean {
            synchronized(this@ArucasList) {
                return this.cursor != this@ArucasList.size
            }
        }

        override fun next(): ClassInstance {
            synchronized(this@ArucasList) {
                val i = this.cursor
                if (i >= this@ArucasList.size) {
                    runtimeError("No such element")
                }
                val data = this@ArucasList.data
                this.last = i
                this.cursor = i + 1
                return data[i]!!
            }
        }

        override fun remove() {
            synchronized(this@ArucasList) {
                if (this.last < 0) {
                    runtimeError("Cannot remove unknown index")
                }
                this@ArucasList.removeAt(this.last)
                if (this.last < this.cursor) {
                    this.cursor--
                }
                this.last = -1
            }
        }

        override fun set(element: ClassInstance) {
            synchronized(this@ArucasList) {
                if (this.last < 0) {
                    runtimeError("Cannot set unknown index")
                }
                this@ArucasList.set(this.last, element)
            }
        }
    }
}